src/ShallowWaterIntegrators.jl

```diff
@@ -0,0 +1,141 @@
+  p = get_trial_fe_basis(P)
+  q = get_fe_basis(Q)
+  n = get_normal_vector(model)
+
+  # explicit step for provisional velocity, uₚ
+  dt1 = dt
+  if leap_frog
+    dt1 = 2.0*dt
+  end
+
+  # 1.1: the mass flux
+  b₁(v)  = ∫(v⋅u₁*h₁)*dΩ
+  rhs1   = assemble_rhs_vector(U, V, b₁(v))
+  op     = AffineFEOperator(U, V, RTMM, rhs1)
+  F      = solve(op)
+  # 1.2: the bernoulli function
+  b₂(q)  = ∫(q*(0.5*u₁⋅u₁ + g*h₁))*dΩ
+  rhs2   = assemble_rhs_vector(P, Q, b₂(q))
+  op     = AffineFEOperator(P, Q, L2MM, rhs2)
+  ϕ      = solve(op)
+  # 1.3: the potential vorticity
+  q₁     = diagnose_potential_vorticity(model, order, Ω, dΩ, qₖ, wₖ, f, h₁, u₁, U, V, R, S)
+  # 1.4: solve for the provisional velocity
+  b₃(v)  = ∫(v⋅uₘ - dt1*(q₁ - τ*u₁⋅∇(q₁))*(v⋅⟂(F,n)))dΩ + ∫(dt1*DIV(v)*ϕ)*dω
+  rhs3   = assemble_rhs_vector(U, V, b₃(v))
+  op     = AffineFEOperator(U, V, RTMM, rhs3)
+  uₚ     = solve(op)
+  # 1.5: solve for the provisional depth
+  b₄(q)  = ∫(q*hₘ)dΩ - ∫(dt1*q*DIV(F))*dω
+  rhs4   = assemble_rhs_vector(P, Q, b₄(q))
+  op     = AffineFEOperator(P, Q, L2MM, rhs4)
+  hₚ     = solve(op)
+
+  # 2.1: the mass flux
+  b₅(v)  = ∫(v⋅u₁*(2.0*h₁ + hₚ)/6.0 + v⋅uₚ*(h₁ + 2.0*hₚ)/6.0)*dΩ
+  rhs5   = assemble_rhs_vector(U, V, b₅(v))
+  op     = AffineFEOperator(U, V, RTMM, rhs5)
+  F      = solve(op)
+  # 2.2: the bernoulli function
+  b₆(q)  = ∫(q*((u₁⋅u₁ + u₁⋅uₚ + uₚ⋅uₚ)/6.0 + 0.5*g*(h₁ + hₚ)))*dΩ
+  rhs6   = assemble_rhs_vector(P, Q, b₆(q))
+  op     = AffineFEOperator(P, Q, L2MM, rhs6)
+  ϕ      = solve(op)
+  # 2.3: the potential vorticity
+  q₂     = diagnose_potential_vorticity(model, order, Ω, dΩ, qₖ, wₖ, f, hₚ, uₚ, U, V, R, S)
+  # 2.4: solve for the final velocity
+  b₇(v)  = ∫(v⋅u₁ - 0.5*dt*(q₁ - τ*u₁⋅∇(q₁) + q₂ - τ*uₚ⋅∇(q₂))*(v⋅⟂(F,n)))dΩ + ∫(dt*DIV(v)*ϕ)*dω
+  rhs7   = assemble_rhs_vector(U, V, b₇(v))
+  op     = AffineFEOperator(U, V, RTMM, rhs7)
+  u₂     = solve(op)
+  # 2.5: solve for the final depth
+  b₈(q)  = ∫(q*h₁)dΩ - ∫(dt*q*DIV(F))*dω
+  rhs8   = assemble_rhs_vector(P, Q, b₈(q))
+  op     = AffineFEOperator(P, Q, L2MM, rhs8)
+  h₂     = solve(op)
+
+  h₂, u₂, ϕ, F
+end
+
+function new_field(A, a)
+  # short cut routine to initialise a new FEFunction 
+  # A: test function
+  # a: dofs to copy into the FEFunction
+  a_dof  = Gridap.FESpaces.get_free_dof_values(a)
+  b_dof  = similar(a_dof)
+  b_dof .= a_dof
+  b      = FEFunction(A, b_dof)
+end
+
+function shallow_water_time_stepper(model, order, Ω, dΩ, dω, qₖ, wₖ, f, g, hn, un, dt, nstep, dump_freq, τ, P, Q, U, V, R, S, method)
+  # Forward integration of the shallow water equations using a supplied method
+
+  # assemble the mass matrices
+  amm(a,b) = ∫(a⋅b)dΩ
+  H1MM = assemble_matrix(amm, R, S)
+  RTMM = assemble_matrix(amm, U, V)
+  L2MM = assemble_matrix(amm, P, Q)
+
+  # initialise the diagnostics arrays
+  mass = zeros(0)
+  vort = zeros(0)
+  kin  = zeros(0)
+  pot  = zeros(0)
+  pow  = zeros(0)
+
+  # first step, no leap frog integration
+  hm1          = new_field(Q, hn)
```

Comment:
I think that `new_field`, though correct, is a bit over-engineered. 

You can do the same more concisely as: 

`FEFunction(Q, copy(get_free_dof_values(hn)))`

so that `new_field` is not required.